File: python/src/reducer.py
from block_utils import multiply_blocks, add_blocks, zero_block

def reducer(key, values, block_size):
    a_blocks = {}
    b_blocks = {}

    for tag, k, block in values:
        if tag == "A":
            a_blocks[k] = block
        else:
            b_blocks[k] = block

    result = zero_block(block_size)

    for k in a_blocks:
        if k in b_blocks:
            partial = multiply_blocks(a_blocks[k], b_blocks[k])
            add_blocks(result, partial)

    return key[0], key[1], result

File: java/src/Reducer.java
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class Reducer {

    public static double[][] reduce(List<BlockValue> values, int blockSize) {
        Map<Integer, double[][]> aBlocks = new HashMap<>();
        Map<Integer, double[][]> bBlocks = new HashMap<>();

        for (BlockValue v : values) {
            if (v.tag.equals("A")) {
                aBlocks.put(v.k, v.block);
            } else {
                bBlocks.put(v.k, v.block);
            }
        }

        double[][] result = BlockUtils.zeroBlock(blockSize);

        for (int k : aBlocks.keySet()) {
            if (bBlocks.containsKey(k)) {
                double[][] partial = BlockUtils.multiply(aBlocks.get(k), bBlocks.get(k));
                BlockUtils.add(result, partial);
            }
        }
        return result;
    }
}
